### DISCUSSION TITLE
Windows builds failing randomly

### DISCUSSION THREAD

**Discussion by asmaloney** [community/community#26933] [Actions]
<p>I have <a href="https://github.com/CloudCompare/CloudCompare/blob/master/.github/workflows/build.yml" rel="nofollow noopener">builds set up</a> for Windows (windows-latest) and macOS - it’s a matrix, so 2 builds run for each.</p>
<p>Sometimes one of the Windows ones fails and cuts off messages like this:</p>
<pre><code>2020-04-19T08:57:16.6185991Z -- Configuring done
2020-04-19T08:57:19.0547176Z -- Generating done
2020-04-19T08:57:19.6715826Z CMake Warning:
2020-04-19T08:57:19.6918171Z ##[error]Process completed with exit code 1.
2020-04-19T08:57:19.6939372Z Cleaning up orphan processes
</code></pre>
<p>…but the other Windows build works. Then I re-run everything using the “re-run” button and they all pass (or they swap and the other Windows one fails this time).</p>
<p>Seems like something is killing the process randomly but don’t have access to any additional info - just the cut-off logs.</p>
<p>This is happening frequently enough that it’s annoying because the macOS ones randomly get “cancelled by user” when nobody has touched it as well :slight_smile:</p>
<p>Windows builds take ~20-25 minutes, so manually re-running all the time is not a good option.</p>
<p>Any idea how to troubleshoot this?</p>

**Comment by Yanjingzhu** [comment]
<p>Could you please <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-runner-diagnostic-logging" rel="nofollow noopener">enable runner diagnostic logging</a> and <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-step-debug-logging" rel="nofollow noopener">enable step debug logging</a> in your workflow? Let’s check whether there are more information in detail. And if you don’t mind, please share a trouble workflow run with debug logs here?  </p>
<p>Please note that, the debug logging will be enabled for all the workflows in the current repository, after you getting a debug log for a failure workflow run, you could disable it.  </p>

**Comment by asmaloney** [comment]
<p>Thank you for the reply. I don’t have admin access to set that up. I’ll have to try to get the organization admin to do it.</p>
<p>In the meantime, <a href="https://github.com/CloudCompare/CloudCompare/runs/603098692?check_suite_focus=true" rel="nofollow noopener">here’s one</a> that failed like this (without the extra debug information). And here is the <a href="https://github.com/CloudCompare/CloudCompare/blob/master/.github/workflows/build.yml" rel="nofollow noopener">workflow YAML</a>.</p>
<p>When I re-run these, they usually pass the first time, but sometimes it takes a couple of tries.</p>
<p>I notice that it’s (often? always?) when it’s doing cmake, so I wonder if it’s a bug in the Windows cmake that’s installed.</p>
<p>I’ll update when I can get that extra information.</p>

**Comment by asmaloney** [comment]
<p>Since I can’t seem to attach anything here in the forum I uploaded the logs <a href="https://github.com/CloudCompare/CloudCompare/issues/1159" rel="nofollow noopener">over here</a>.</p>
<p>I took a look through them and can’t see anything obvious.</p>
<p>The one that failed this time is <strong>Windows Latest MSVC SCALAR_DOUBLE=OFF</strong> in step 7_Configure CMake.txt (sometimes it’s <strong>Windows Latest MSVC SCALAR_DOUBLE=ON</strong> that fails though).</p>
<pre><code>...
2020-04-21T22:54:16.1043114Z -- QHULL found (include: C:/Miniconda3/envs/ccenv/Library/include, lib: optimized;C:/Miniconda3/envs/ccenv/Library/lib/qhullstatic.lib;debug;C:/Miniconda3/envs/ccenv/Library/lib/qhullstatic.lib)
2020-04-21T22:54:16.1043705Z -- looking for PCL_COMMON
2020-04-21T22:54:16.1058916Z -- Found PCL_COMMON: C:/Miniconda3/envs/ccenv/Library/lib/pcl_common_release.lib
2020-04-21T22:54:16.8933755Z CMake Warning (dev) at C:/Program Files/CMake/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:272 (message):
2020-04-21T22:54:19.1474864Z ##[error]Process completed with exit code 1.
2020-04-21T22:54:19.1478756Z ##[debug]Finishing: Configure CMake
</code></pre>
<p>I don’t see anything in the additional Runner and Worker logs that looks strange, but I don’t know what I’m looking for exactly.</p>

**Comment by Yanjingzhu** [comment]
<p>I have forked your repo to my org. I met the same error as yours. I am trying to involve github engineering team to further investigate this issue. There might be some time delay. Appreciate your patience.</p>

**Comment by asmaloney** [comment]
<p>Great - thank you for your time.</p>
<p>This was a very large time sink yesterday as I had to re-run dozens of times.</p>
<p>If you (or they) have any suggestions on how to reduce the build time on Windows (~25 minutes) I’d love to hear that as well :slight_smile:</p>

**Comment by asmaloney** [comment]
<p>I have another run that keeps failing - I’ve tried it at least a dozen times today. I know it should pass because it passed the PR request stage, was merged, and now the merged CI keeps failing :slight_smile:</p>
<p>I turned on the logging and found this in one of the “worker” logs - maybe it will help?</p>
<pre><code>[2020-04-23 17:14:49Z INFO HostContext] No proxy settings were found based on environmental variables (http_proxy/https_proxy/HTTP_PROXY/HTTPS_PROXY)
[2020-04-23 17:14:49Z INFO HostContext] Well known directory 'Bin': 'C:\runners\2.169.1\bin'
[2020-04-23 17:14:49Z INFO HostContext] Well known directory 'Root': 'C:\runners\2.169.1'
[2020-04-23 17:14:49Z INFO HostContext] Well known config file 'Credentials': 'C:\runners\2.169.1\.credentials'
[2020-04-23 17:14:49Z INFO Worker] Version: 2.169.1
[2020-04-23 17:14:49Z INFO Worker] Commit: 4f840647b375b65a4efa361abc1b84f783442de6
[2020-04-23 17:14:49Z INFO Worker] Culture: en-US
[2020-04-23 17:14:49Z INFO Worker] UI Culture: en-US
[2020-04-23 17:14:49Z ERR Worker] System.ArgumentException: Handle has been disposed or is invalid. (Parameter 'pipeHandleAsString')
   at System.IO.Pipes.AnonymousPipeClientStream..ctor(PipeDirection direction, String pipeHandleAsString)
   at GitHub.Runner.Common.ProcessChannel.StartClient(String pipeNameInput, String pipeNameOutput)
   at GitHub.Runner.Worker.Worker.RunAsync(String pipeIn, String pipeOut)
   at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args)
</code></pre>

**Comment by Yanjingzhu** [comment]
<p>Thank you for looking deep into this issue . I am still trying to contact to senior engineer , your patience would be appreciated. </p>

**Comment by asmaloney** [comment]
<p>Thank you.</p>
<p>We have had to re-run these things dozens of times (and you can’t just re-run the ones that fail - it’s all-or-nothing), so not only has this been very frustrating, it has been a major time sink and a waste of build resources at GitHub.</p>
<p>Combine this with the macOS runners randomly deciding someone cancelled them, and it shakes my confidence in Actions as a CI system. It kind of goes without saying that we need to be able to trust this and rely on it to be rock-solid.</p>

**Comment by AlenaSviridenko** [comment]
<p>Hi <a class="mention" href="/asmaloney">@asmaloney</a> !</p>
<p>Thank you for your patience.</p>
<p>Could you please check the investigation results below and see if it helps to solve the issue?</p>
<p>It turned out that by default CMake is not displaying its internal errors, and in order to find a reason of the fail we changed the workflow to show the CMake logs and saw there’s a compilation error from the C++ build itself</p>
<pre><code>CMakeError.log content
Performing C++ SOURCE FILE Test COMPILER_HAS_DEPRECATED_ATTR failed with the following output:
Change Dir: D:/a/CloudCompare/CloudCompare/build/CMakeFiles/CMakeTmp
Run Build Command(s):C:/Miniconda3/envs/ccenv/Library/bin/ninja.exe cmTC_144a7 &amp;&amp; [1/2] Building CXX object CMakeFiles\cmTC_144a7.dir\src.cxx.obj
FAILED: CMakeFiles/cmTC_144a7.dir/src.cxx.obj
C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1425~1.286\bin\Hostx64\x64\cl.exe /nologo /TP /DWIN32 /D_WINDOWS /W3 /GR /EHsc -DCOMPILER_HAS_DEPRECATED_ATTR /MDd /Zi /Ob0 /Od /RTC1 -std:c++14 /showIncludes /FoCMakeFiles\cmTC_144a7.dir\src.cxx.obj /FdCMakeFiles\cmTC_144a7.dir\ /FS -c src.cxx
src.cxx(1): error C2065: ' __deprecated__': undeclared identifier
src.cxx(1): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
src.cxx(1): error C2448: ' __attribute__': function-style initializer appears to be a function definition
ninja: build stopped: subcommand failed.
</code></pre>
<p>The diff to show logs: <a href="https://github.com/dsame/CloudCompare/commit/e8248d498ba8835cef2e512532349442a18b9711" rel="nofollow noopener">dsame/CloudCompare@e8248d4</a> (idea is to surround cmake… with try…catch and output all the logs)<br>
The sample build: <a href="https://github.com/dsame/CloudCompare/runs/645255014?check_suite_focus=true" rel="nofollow noopener">https://github.com/dsame/CloudCompare/runs/645255014?check_suite_focus=true</a></p>

**Comment by asmaloney** [comment]
<p>Thanks for looking into it Alena.</p>
<p>…but the build you’re showing didn’t fail. What you’re showing here (and in the log) looks like the standard way CMake tests for compiler attributes. In that case it’s supposed to fail.</p>
<p>(The code it points at is cmake’s code for testing compilers for specific atttributes - not our code.)</p>
<p>The build works on a real Windows machine (the rest of the team are using them).</p>
<p>Right now with master it usually fails several times before it builds properly. Here’s a more recent example:</p>
<p><a href="https://github.com/CloudCompare/CloudCompare/runs/618602537" rel="nofollow noopener">https://github.com/CloudCompare/CloudCompare/runs/618602537</a></p>
<p>If I keep re-running that one it will pass eventually.</p>
<p>So maybe try re-running it until it fails?</p>
<p>Did you see my comment above about the worker logs and pipes? That seems more likely to be an issue.</p>

**Comment by AlenaSviridenko** [comment]
<p>Your comment makes sense. We were actually able to reproduce the issue from your forked repo, and the only errors we were able to get are the ones in the build I provided in the answer above (that build I attached did not fail because of applied logic in try…catch):</p>
<pre><code>CheckIncludeFile.c(1): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory
ninja: build stopped: subcommand failed.
</code></pre>
<p>Could you please add the suggested patch to your code, and provide the logs you will get? it will help us to understand what is happening in your build. </p>

**Comment by asmaloney** [comment]
<p>&gt; did not fail because of applied logic in try…catch</p>
<p>I’m sorry, but that’s not the case. As I mentioned above, the things you are showing as errors here will not cause the build to fail. They are normal - it’s the way that cmake determines what’s available on the machine.</p>
<p>So I don’t think you have reproduced the error with the try…catch yet. It fails in the cmake config process, so it never starts the build step. What about the other error I mentioned with the runner? Again - it seems a likely candidate.</p>
<p>I will see what else I can do at my end, but please try to reproduce it again. The symptom is (I think) that cmake is trying to produce a warning but instead something (runner?) produces an error which kills the build. It doesn’t happen all the time, but it happens very frequently. I’ve had to run them 16 or 17 times in some cases to get it to pass.</p>

**Comment by dsame** [comment]
<p>The reason builds fails is either CMake or another tool invoked by CMake writes to stderr. PowerShell triggers terminating exception then.</p>
<p>The reason build fails randomly is a race condition we did not identified exactly becaues the build system is too complex and not obvious. </p>
<p>In order to identify the exact the problem of CMake fail i suggest you to apply the code of this PR: <a href="https://github.com/CloudCompare/CloudCompare/pull/1232" rel="nofollow noopener">https://github.com/CloudCompare/CloudCompare/pull/1232</a> which intended to intercept the stderr stream and to do not let it miss the logs and PowerShell output. </p>
<p>With this changes the CMake Configure will not break on the warnings but print them to the build log, so you can analyze them and find out the unexpected behaviour. </p>
<p>The sample build: <a href="https://github.com/dsame/CloudCompare/runs/694200554?check_suite_focus=true" rel="nofollow noopener">https://github.com/dsame/CloudCompare/runs/694200554?check_suite_focus=true</a></p>
<p>The messages cause PowerShell to fail becaues of CMake writes to stder are marked with “ERR” prefix:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/80e3d11c1918fb5f2bf89132a418ae6c375160e7.png" data-download-href="/uploads/short-url/iodd78NOFFIfT9owgKSkBY1xkAn.png?dl=1" title="2020-05-21_574.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/245540/181100862-28b672ab-a5f6-4cfd-8330-5c13635e98c7.png" alt="2020-05-21_574.png" data-base62-sha1="iodd78NOFFIfT9owgKSkBY1xkAn" width="690" height="373" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/80e3d11c1918fb5f2bf89132a418ae6c375160e7_2_690x373.png, https://aws1.discourse-cdn.com/github/optimized/1X/80e3d11c1918fb5f2bf89132a418ae6c375160e7_2_1035x559.png 1.5x, https://aws1.discourse-cdn.com/github/original/1X/80e3d11c1918fb5f2bf89132a418ae6c375160e7.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/80e3d11c1918fb5f2bf89132a418ae6c375160e7_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">2020-05-21_574.png</span><span class="informations">1193×645 388 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Please, review the suggested PR and confirm it can help effectively to solve the problem</p>

**Comment by dsame** [comment]
<p>There’s a slightly modified build.yml which fails in case if there is any sdterr otput from the build. It seems to be too strict because but you might find it usefull</p>
<p><a href="https://github.com/dsame/CloudCompare/blob/9888fc2a5e26d7c8bb0059f79f6d2bd1db881921/.github/workflows/build.yml" rel="nofollow noopener">https://github.com/dsame/CloudCompare/blob/9888fc2a5e26d7c8bb0059f79f6d2bd1db881921/.github/workflows/build.yml</a></p>
<p>Please try to apply it your build in order to see the stderr output that was lost in your build and find out the differences in success and failed builds. </p>
<p>Alternatively, as the recent Windows images have bash installed i’d advise to switch to it from PowerShell:<br>
<a href="https://github.com/CloudCompare/CloudCompare/compare/CloudCompare:master...dsame:dsame/use-cmd?expand=1" rel="nofollow noopener">https://github.com/CloudCompare/CloudCompare/compare/CloudCompare:master…dsame:dsame/use-cmd?expand=1</a></p>
<p>With the bash you should not have the problem with CMake output to stderr</p>

**Comment by asmaloney** [comment]
<p>I like the idea of switching to bash since I’m completely unfamiliar with PowerShell anyways. It seems like the easiest option. I’ll do that and report back.</p>
<p>Thanks for your help Sergey!</p>

**Comment by asmaloney** [comment, marked as answer]
<p>After switching from <em>PowerShell</em> to <em>bash</em> as suggested by <a class="mention" href="/dsame">@dsame</a>, we have not had any of the random failures on Windows we were seeing before.</p>
<p><strong>Solution:</strong> Don’t use PowerShell.</p>
<p>Thanks <a class="mention" href="/dsame">@dsame</a>, <a class="mention" href="/alenasviridenko">@alenasviridenko</a> and <a class="mention" href="/yanjingzhu">@yanjingzhu</a> for looking into it.</p>
